<a href="https://colab.research.google.com/github/chiro2099/Random_Forest_Project/blob/main/KGP_intern_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [154]:
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd

In [155]:
df=pd.read_excel('/content/kgp_data_modified.xlsx')

In [156]:
df.head()

Sl. No.   Gender  Age Course year  CGPA Depression Anxiety Panic attack  \
0         1    Male   19    CSE  1st  6.88         NO     YES           NO   
1         2  Female   20    CSE  1st  7.49         NO      NO           NO   
2         3   Male    18    CSE  1st  8.66         NO      NO           NO   
3         4   Male    19    CSE  1st  5.74        YES      NO           NO   
4         5    Male   18    CSE  1st  6.86         NO     YES          YES   

  Specialist treatment  
0                   NO  
1                   NO  
2                   NO  
3                   NO  
4                   NO

In [157]:
df.columns

Index(['Sl. No. ', 'Gender', 'Age', 'Course', 'year', 'CGPA', 'Depression',
       'Anxiety', 'Panic attack', 'Specialist treatment'],
      dtype='object')

In [158]:
df.shape

(151, 10)

In [159]:
df.isnull().sum().sum()

0

In [160]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()


In [161]:
df.Gender.unique()

array(['Male', 'Female', 'Male ', 'FeMale', 'Female '], dtype=object)

In [162]:
df

Sl. No.   Gender  Age Course year  CGPA Depression Anxiety Panic attack  \
0           1    Male   19    CSE  1st  6.88         NO     YES           NO   
1           2  Female   20    CSE  1st  7.49         NO      NO           NO   
2           3   Male    18    CSE  1st  8.66         NO      NO           NO   
3           4   Male    19    CSE  1st  5.74        YES      NO           NO   
4           5    Male   18    CSE  1st  6.86         NO     YES          YES   
..        ...     ...  ...    ...  ...   ...        ...     ...          ...   
146       146   Male    22    EEE  3rd  8.07         NO      NO           NO   
147       147    Male   22    EEE  3rd  7.80         NO      NO           NO   
148       148  Female   20    EEE  3rd  8.18         NO      NO           NO   
149       149  Female   20    EEE  3rd  7.04         NO      NO           NO   
150       150    Male   21    EEE  3rd  8.15         NO      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 10 columns]

In [163]:
corrections={'FeMale': 'Female', 'feamale': 'Female ', 'Female ': 'Female'}
corrections2={'Male ':'Male'}
df['Gender'] = df['Gender'].replace(corrections)
df['Gender'] = df['Gender'].replace(corrections2)

In [164]:
df

Sl. No.   Gender  Age Course year  CGPA Depression Anxiety Panic attack  \
0           1    Male   19    CSE  1st  6.88         NO     YES           NO   
1           2  Female   20    CSE  1st  7.49         NO      NO           NO   
2           3    Male   18    CSE  1st  8.66         NO      NO           NO   
3           4    Male   19    CSE  1st  5.74        YES      NO           NO   
4           5    Male   18    CSE  1st  6.86         NO     YES          YES   
..        ...     ...  ...    ...  ...   ...        ...     ...          ...   
146       146    Male   22    EEE  3rd  8.07         NO      NO           NO   
147       147    Male   22    EEE  3rd  7.80         NO      NO           NO   
148       148  Female   20    EEE  3rd  8.18         NO      NO           NO   
149       149  Female   20    EEE  3rd  7.04         NO      NO           NO   
150       150    Male   21    EEE  3rd  8.15         NO      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 10 columns]

In [165]:
df.Gender.unique()

array(['Male', 'Female'], dtype=object)

In [166]:
df=df.drop(['Sl. No. '],axis='columns')

In [167]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()


In [168]:
df['Gender']=encoder.fit_transform(df['Gender'])

In [169]:
df

Gender  Age Course year  CGPA Depression Anxiety Panic attack  \
0         1   19    CSE  1st  6.88         NO     YES           NO   
1         0   20    CSE  1st  7.49         NO      NO           NO   
2         1   18    CSE  1st  8.66         NO      NO           NO   
3         1   19    CSE  1st  5.74        YES      NO           NO   
4         1   18    CSE  1st  6.86         NO     YES          YES   
..      ...  ...    ...  ...   ...        ...     ...          ...   
146       1   22    EEE  3rd  8.07         NO      NO           NO   
147       1   22    EEE  3rd  7.80         NO      NO           NO   
148       0   20    EEE  3rd  8.18         NO      NO           NO   
149       0   20    EEE  3rd  7.04         NO      NO           NO   
150       1   21    EEE  3rd  8.15         NO      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 9 columns]

In [170]:
df.Course.unique()

array(['CSE', 'ECE', 'EEE'], dtype=object)

In [171]:
from sklearn.preprocessing import LabelEncoder
encoder_another=LabelEncoder()

In [172]:
df['Course']=encoder_another.fit_transform(df['Course'])

In [173]:
df

Gender  Age  Course year  CGPA Depression Anxiety Panic attack  \
0         1   19       0  1st  6.88         NO     YES           NO   
1         0   20       0  1st  7.49         NO      NO           NO   
2         1   18       0  1st  8.66         NO      NO           NO   
3         1   19       0  1st  5.74        YES      NO           NO   
4         1   18       0  1st  6.86         NO     YES          YES   
..      ...  ...     ...  ...   ...        ...     ...          ...   
146       1   22       2  3rd  8.07         NO      NO           NO   
147       1   22       2  3rd  7.80         NO      NO           NO   
148       0   20       2  3rd  8.18         NO      NO           NO   
149       0   20       2  3rd  7.04         NO      NO           NO   
150       1   21       2  3rd  8.15         NO      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 9 columns]

In [174]:
df.year.unique()

array(['1st', '2nd', '3rd', '4th'], dtype=object)

In [175]:
from sklearn.preprocessing import LabelEncoder
encoder_year=LabelEncoder()

In [176]:
df['year']=encoder_year.fit_transform(df['year'])

In [177]:
df.head()

Gender  Age  Course  year  CGPA Depression Anxiety Panic attack  \
0       1   19       0     0  6.88         NO     YES           NO   
1       0   20       0     0  7.49         NO      NO           NO   
2       1   18       0     0  8.66         NO      NO           NO   
3       1   19       0     0  5.74        YES      NO           NO   
4       1   18       0     0  6.86         NO     YES          YES   

  Specialist treatment  
0                   NO  
1                   NO  
2                   NO  
3                   NO  
4                   NO

In [178]:
df

Gender  Age  Course  year  CGPA Depression Anxiety Panic attack  \
0         1   19       0     0  6.88         NO     YES           NO   
1         0   20       0     0  7.49         NO      NO           NO   
2         1   18       0     0  8.66         NO      NO           NO   
3         1   19       0     0  5.74        YES      NO           NO   
4         1   18       0     0  6.86         NO     YES          YES   
..      ...  ...     ...   ...   ...        ...     ...          ...   
146       1   22       2     2  8.07         NO      NO           NO   
147       1   22       2     2  7.80         NO      NO           NO   
148       0   20       2     2  8.18         NO      NO           NO   
149       0   20       2     2  7.04         NO      NO           NO   
150       1   21       2     2  8.15         NO      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 9 columns]

In [179]:
len(df[df['year']==1])

40

**Predicting Occurrence of Depression**

---



In [180]:
df.Depression.unique()

array(['NO', 'YES', 'NO '], dtype=object)

In [181]:
correctio={'NO ':'NO'}
df['Depression']=df['Depression'].replace(correctio)

In [182]:
df.Depression.unique()

array(['NO', 'YES'], dtype=object)

In [183]:
from sklearn.preprocessing import LabelEncoder
encoder_dep=LabelEncoder()

In [184]:
df['Depression']=encoder_dep.fit_transform(df['Depression'])

In [186]:
df.Depression.unique()

array([0, 1])

In [188]:
df.columns

Index(['Gender', 'Age', 'Course', 'year', 'CGPA', 'Depression', 'Anxiety',
       'Panic attack', 'Specialist treatment'],
      dtype='object')

In [192]:
df_new=df.drop(['Depression','Panic attack','Anxiety','Specialist treatment'],axis='columns')

In [193]:
df_new

Gender  Age  Course  year  CGPA
0         1   19       0     0  6.88
1         0   20       0     0  7.49
2         1   18       0     0  8.66
3         1   19       0     0  5.74
4         1   18       0     0  6.86
..      ...  ...     ...   ...   ...
146       1   22       2     2  8.07
147       1   22       2     2  7.80
148       0   20       2     2  8.18
149       0   20       2     2  7.04
150       1   21       2     2  8.15

[151 rows x 5 columns]

In [194]:
x=df_new
y=df['Depression']

In [195]:
from sklearn.model_selection import train_test_split

In [196]:
x_train, x_test, y_train, y_test =train_test_split(x,y, test_size=0.2)

In [197]:
from xgboost import XGBClassifier
model= XGBClassifier()

In [199]:
model.fit(x_train,y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

In [200]:
model.predict(x_test)

array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0])

In [201]:
model.score(x_test,y_test)

0.8387096774193549

Obtained a AUC-ROC score of 0.839 using XGBoost

**Predicting probability of seeking specialist treatment**

---



In [202]:
df

Gender  Age  Course  year  CGPA  Depression Anxiety Panic attack  \
0         1   19       0     0  6.88           0     YES           NO   
1         0   20       0     0  7.49           0      NO           NO   
2         1   18       0     0  8.66           0      NO           NO   
3         1   19       0     0  5.74           1      NO           NO   
4         1   18       0     0  6.86           0     YES          YES   
..      ...  ...     ...   ...   ...         ...     ...          ...   
146       1   22       2     2  8.07           0      NO           NO   
147       1   22       2     2  7.80           0      NO           NO   
148       0   20       2     2  8.18           0      NO           NO   
149       0   20       2     2  7.04           0      NO           NO   
150       1   21       2     2  8.15           0      NO           NO   

    Specialist treatment  
0                     NO  
1                     NO  
2                     NO  
3                     NO  
4                     NO  
..                   ...  
146                   NO  
147                   NO  
148                   NO  
149                   NO  
150                   NO  

[151 rows x 9 columns]

In [205]:
df.columns

Index(['Gender', 'Age', 'Course', 'year', 'CGPA', 'Depression', 'Anxiety',
       'Panic attack', 'Specialist treatment'],
      dtype='object')

In [206]:
correctio={'NO ':'NO'}
df['Anxiety']=df['Anxiety'].replace(correctio)
df['Panic attack']=df['Panic attack'].replace(correctio)

In [204]:
df['Anxiety'].unique()

array(['YES', 'NO'], dtype=object)

In [207]:
df['Panic attack'].unique()

array(['NO', 'YES'], dtype=object)

In [208]:
df['Specialist treatment'].unique()

array(['NO', 'YES'], dtype=object)

In [212]:
from sklearn.preprocessing import LabelEncoder
encoder_nuevo=LabelEncoder()


In [214]:
df['Anxiety']=encoder_nuevo.fit_transform(df['Anxiety'])
df['Panic attack']=encoder_nuevo.fit_transform(df['Panic attack'])
df['Specialist treatment']=encoder_nuevo.fit_transform(df['Specialist treatment'])

In [215]:
df

Gender  Age  Course  year  CGPA  Depression  Anxiety  Panic attack  \
0         1   19       0     0  6.88           0        1             0   
1         0   20       0     0  7.49           0        0             0   
2         1   18       0     0  8.66           0        0             0   
3         1   19       0     0  5.74           1        0             0   
4         1   18       0     0  6.86           0        1             1   
..      ...  ...     ...   ...   ...         ...      ...           ...   
146       1   22       2     2  8.07           0        0             0   
147       1   22       2     2  7.80           0        0             0   
148       0   20       2     2  8.18           0        0             0   
149       0   20       2     2  7.04           0        0             0   
150       1   21       2     2  8.15           0        0             0   

     Specialist treatment  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
..                    ...  
146                     0  
147                     0  
148                     0  
149                     0  
150                     0  

[151 rows x 9 columns]

In [216]:
x2=df.drop(['Specialist treatment'],axis='columns')

In [217]:
y2=df['Specialist treatment']

In [218]:
x2_train,x2_test, y2_train, y2_test=train_test_split(x2, y2, test_size=0.4)

In [219]:
model2=XGBClassifier()

In [220]:
model2.fit(x2_train, y2_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

In [223]:
model2.predict(x2_test)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [225]:
model2.score(x2_test,y2_test)

0.9180327868852459

**Obtained an ROC-AUC score of 0.918 using XGBoost**